src/components/i18n/FormattedCompMessage.js

```diff
@@ -0,0 +1,205 @@
+                }
+            } else {
+                source = composition.compose();
+            }
+
+            this.state = {
+                id,
+                source,
+                composition,
+            };
+        }
+    }
+
+    /**
+     * Search for any Plural elements in the children, and
+     * then construct the English source string in the correct
+     * format for react-intl to use for pluralization
+     * @param {React.Element} children the children of this node
+     * @return {string} the composed plural string
+     */
+    composePluralString(children) {
+        const categories = {};
+        React.Children.forEach(children, child => {
+            if (typeof child === 'object' && React.isValidElement(child) && child.type.name === 'Plural') {
+                const childComposition = new Composition(child.props.children);
+                categories[child.props.category] = childComposition.compose();
+            }
+        });
+        if (!categories.one || !categories.other) {
+            if (isDevEnvironment()) {
+                throw new Error(
+                    'Cannot use count prop on a FormattedCompMessage component without giving both a "one" and "other" Plural component in the children.',
+                );
+            }
+        }
+        // add these to the string in a particular order so that
+        // we always end up with the same string regardless of
+        // the order that the Plural elements were specified in
+        // the source code
+        const categoriesString = [
+            CATEGORY_ZERO,
+            CATEGORY_ONE,
+            CATEGORY_TWO,
+            CATEGORY_FEW,
+            CATEGORY_MANY,
+            CATEGORY_OTHER,
+        ]
+            .map(category => (categories[category] ? ` ${category} {${categories[category]}}` : ''))
+            .join('');
+
+        // see the intl-messageformat project for an explanation of this syntax
+        return `{count, plural,${categoriesString}}`;
+    }
+
+    render() {
+        const { count, tagName, intl, description } = this.props;
+        const { composition, id, source } = this.state;
+        const values = {};
+        if (typeof count === 'number') {
+            // make sure intl.formatMessage switches properly on the count
+            values.count = count;
+        }
+
+        // react-intl will do the correct plurals if necessary
+        const translation = intl.formatMessage(
+            {
+                id,
+                defaultMessage: source,
+                description,
+            },
+            values,
+        );
+
+        // always wrap the translated string in a tag to contain everything
+        // and to give us a spot to record the id
+        return React.createElement(
+            tagName || 'span',
+            {
+                key: id,
+                'x-resource-id': id,
```

Comment:
Can you add a comment explaining the purpose of this attribute?

Reply:
This documents the resource id of the string in mojito. This can be used for a number of purposes. It is one of the most important parts of this whole component from my point-of-view as a moji guy!

When the strings are tagged with their id like that, we can make a browser plugin that will connect strings in the UI directly to mojito. (Right-click > View This String in Mojito). This allows us to do in-context linguistic review of the translations, as well as updating the strings on the fly. In fact the linguistic reviewers can submit the translation fixes directly into mojito and then preview their proposed fix in the browser on the fly! (Queue the singing angels!)

It also allows us to figure out where a string is used in the UI so that when we take screen shots for the translators, we can connect those screen shots to particular strings in mojito.

It is also very useful for debugging translation problems, as it tells us where each UI string comes from. (Which Box product, which repo, which string, etc.) Many times, especially for short strings, there are multiple places in the code where the exact same string is used. Which one do we fix? Probably not all of them... context is everything! Currently, we have to spend a lot of time sleuthing around in multiple git repos to find the exact string, and guess which one we need to update in mojito. It gets even worse when a jira ticket comes in with a screen shot in a different language telling us to fix the string to something else. How do you even type that text in the search box in mojito? (Do you know how to type that string in the screen shot in Russian, Chinese, or Japanese? I know I don't!) If the id is right there in the code, then it is easy to find and update the string in mojito.

Later, the same id can be used to implement the text A/B testing framework that the personalization/growth team is interested in building. The id will allow a product manager to right-click on any text in the UI and a browser plugin will find that id and throw up a dialog box that allows the PM to enter test variations for the A/B test, pick some criteria, and then launch the test. (Again using the right browser plugin.)

For everyone else, the spans are not seen in the UI and resource ids are not visible.

Reply:
This is great context. I didn't realize we had a separate plugin to assist with translations. Can you add a shorter version of this information to the source file? That way anyone who's perusing this component in the future will be able to see its purpose immediately.

Reply:
I still think having a brief in-code comment to explain the purpose of this attribute would be prudent.